src/sentry/workflow_engine/endpoints/validators/metric_alert_detector.py

```diff
@@ -38,3 +43,75 @@ def validate(self, attrs):
         if len(conditions) > 2:
             raise serializers.ValidationError("Too many conditions")
         return attrs
+
+    def update_data_conditions(self, instance, data_conditions):
+        """
+        Update the data condition if it already exists, create one if it does not
+        """
+        if instance.workflow_condition_group:
+            try:
+                data_condition_group = DataConditionGroup.objects.get(
+                    id=instance.workflow_condition_group.id
+                )
+            except DataConditionGroup.DoesNotExist:
+                raise serializers.ValidationError("DataConditionGroup not found, can't update")
+        # TODO make one if it doesn't exist and data is passed?
+
+        for data_condition in data_conditions:
+            try:
+                current_data_condition = DataCondition.objects.get(
+                    id=data_condition.get("id"), condition_group=data_condition_group
+                )
+            except DataCondition.DoesNotExist:
+                current_data_condition = None
+
+            if current_data_condition:
+                current_data_condition.update(**data_condition)
+            else:
+                DataCondition.objects.create(
+                    type=data_condition.get("type"),
+                    comparison=data_condition.get("comparison"),
+                    condition_result=data_condition.get("condition_result"),
+                    condition_group=data_condition_group,
+                )
+        return data_condition_group
+
+    def update_data_source(self, instance, data_source):
+        try:
+            source_instance = DataSource.objects.get(detector=instance)
+        except DataSource.DoesNotExist:
```

Comment:
does the data source need to be created if you add it to an existing detector?

Reply:
I am not sure if that scenario would be possible in the UI, I believe you're always creating both a data source and a detector at the same time